Improve include/exclude search in workspace

alvsan09 opened the pull request:
The following descriptions apply to the user input for the fields
'Files to include' and 'Files to exclude' for the 'Search in
workspace' feature.

Enhancements:
* Handle provided relative paths e.g. ./README.md, README.md to only
  apply the specified file.
* Includes are limited to specific folder e.g. './test' should
  not include './dir/path/test'.
* Input strings (i.e. not resolving to relative paths) are converted
  to two globs in order to include results of a resulting file and
  include files under a resulting folder i.e. resolved globs are
  **`**/string`** and **`**/string/*`**.
  E.g. input file '.snaphot' shall include a file '.snapshot' or if
  it's a folder it should apply to all the files underneath it.
* 'Files to include' can reference folders outside the workspace

Fixes: #8469

**The search strategy offered by the search in workspace server
(RipgrepSearchInWorkspaceServer.search) is changed as follows:**

By default, sets the search directories for the string WHAT to the provided ROOTURIS directories
and returns the assigned search id.

The include / exclude (options in SearchInWorkspaceOptions) are lists of patterns for files to
include / exclude during search (glob characters are allowed).

include patterns successfully recognized as absolute paths will override the default search and set
the search directories to the ones provided as includes.
Relative paths are allowed, the application will attempt to translate them to valid absolute paths
based on the applicable search directories.

Signed-off-by: Alvaro Sanchez-Leon <alvaro.sanchez-leon@ericsson.com>

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
For strings provided for files to include / exclude
* Checks if the 'include' strings provided are valid paths to the available root folders and if so convert them to glob patterns
* If the include strings are valid paths, replace the root paths with the resulting paths to be consumed by the 'ripgrep' executable.
* Consider each string not resolving to sub-folders as a pattern to either a file or folder, therefore generating two glob entries for each.

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
4 new unit test cases provided

Manual testing can be performed as follows:

Using theia repo as an example directory structure:

1. Search to a relative file path of workspace root(s)
    - Search pattern: cloud  (produces a handful of matches)
    - Enter 'Files to Include' text to: ./README.md  (verify that only one file is considered)
2. 'Files to include' can refer to files
    - Search pattern: example
    - Enter 'Files to include' to: `.eslintcache` (verify that multiple matches are reported)
3. 'Files to include' can refer to folders
    - Search pattern: example
    - Enter 'Files to include' to: test (verify that multiple matches are reported)
4. Searching limited to a given subfolder
    -  Search pattern: example
    - Enter 'Files to include' to: **/doc/* (note that matches in two different folders are found i.e. 'doc' and 'packages/plugin-ext/doc'
    - Replace 'Files to include' to doc (note that this now resolves matches for a single folder i.e. 'doc')
5. From the explorer view select a file or folder and from the context menu select 'Find in folder', press enter, and verify the results are as expected.
6. Using the 'Files to include' box, enter absolute paths to external folders, validate expected results.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -20,6 +20,11 @@ import { FileUri } from '@theia/core/lib/node/file-uri';
 import { inject, injectable } from '@theia/core/shared/inversify';
 import { SearchInWorkspaceServer, SearchInWorkspaceOptions, SearchInWorkspaceResult, SearchInWorkspaceClient, LinePreview } from '../common/search-in-workspace-interface';
+import * as fs from '@theia/core/shared/fs-extra';
+import path = require('path');
```
Can likely be converted:

```suggestion
import * as path from 'path';
```

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +98,91 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
-            args.push('--fixed-strings');
-            args.push('--');
+            args.add('--fixed-strings');
+            args.add('--');
```
I noticed the following as well when we execute `rg`, do you know why we need the trailing `--`?

Ex (searching for _about_):

```
root INFO processOptions: command = /home/evinfug/workspaces/theia/node_modules/vscode-ripgrep/bin/rg
root INFO processOptions: args = 
--hidden
--json
--ignore-case
--max-filesize=20M
--glob=!**/**/.git
--glob=!**/**/.svn
--glob=!**/**/.hg
--glob=!**/**/CVS
--glob=!**/**/.DS_Store
--fixed-strings
--
about
/home/evinfug/workspaces/theia
```

vince-fugnitto replied:
@alvsan09 thank you for following up! 
Given that description shouldn't `--` be added in all cases and not solely in the `else`?

alvsan09 replied:
I tried it, and it unfortunately break the behavior as several unit test fail

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +98,91 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
 
+    protected pathToGlobs(inputPath: string, rootFolder: string, exclude: boolean): string[] {
+        const prefixPattern = (inputPattern: string): string => {
+            const globalCommandArgument = '--glob=';
```
I think we mean `glob` here and not `global`:

```suggestion
            const globCommandArgument = '--glob=';
```

alvsan09 replied:
Done, Thanks !!

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +98,91 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+            return `${globalCommandArgument}${excludeChar}${updatedPattern}`;
+        };
+
+        const toGlobPattern = pathToGlobPattern({
+            extensions: [],
+            cwd: rootFolder
+        });
+
+        const pattern = isGlob(inputPath) ? inputPath : toGlobPattern(inputPath);
```
We can likely save computation by doing it directly when needed:

```suggestion
const pattern = isGlob(inputPath)
  ? inputPath
  : pathToGlobPattern({ extensions: [], cwd: rootFolder })(inputPath);
```

alvsan09 replied:
Done, Thanks !

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -290,6 +339,56 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+    }
+
+    protected includeSubDirectoriesToPath(pattern: string): string {
+        const parsedObj = path.parse(pattern);
```
`parsedPath`? 

```suggestion
        const parsedPath = path.parse(pattern);
```

alvsan09 replied:
Done, Thanks !

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -290,6 +339,56 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+        return searchPaths.length > 0 ? searchPaths : rootPaths;
+    }
+
+    protected includeSubDirectoriesToPath(pattern: string): string {
```
Can you document what this method is attempting to achieve?
Can it not be generalized rather than only for `include`?

It might also help to unit-test it to verify it is behaving as intended under different use-cases.

alvsan09 replied:
Done, the newest patch set reduced the size of the methods added, I have added some doc info and additional test cases were created.

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +156,94 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
-    protected getArgs(options?: SearchInWorkspaceOptions): string[] {
-        const args = ['--hidden', '--json'];
-        args.push(options && options.matchCase ? '--case-sensitive' : '--ignore-case');
+    protected getArgs(rootPaths: string[], options?: SearchInWorkspaceOptions): string[] {
```
We should reflect the breaking change in the `changelog`.

alvsan09 replied:
Done

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.slow-spec.ts:
```diff
@@ -848,3 +930,109 @@ describe('ripgrep-search-in-workspace-server', function (): void {
     });
 });
+
+describe('ripgrep-search-in-workspace-server-matchPatternToPathResult', function (): void {
```
It would be sufficient to name it the following way:

```suggestion
describe('#matchPatternToPathResult', function (): void {
```

alvsan09 replied:
Done

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.slow-spec.ts:
```diff
@@ -848,3 +930,109 @@ describe('ripgrep-search-in-workspace-server', function (): void {
+
+describe('ripgrep-search-in-workspace-server-matchPatternToPathResult', function (): void {
+    this.timeout(10000);
+    it('Resolve pattern to path for filename', function (): void {
```
We should follow the format of existing tests:

```suggestion
    it('should resolve pattern to path for filename', function (): void {
```

The same is true for all other tests.

alvsan09 replied:
Done

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.slow-spec.ts:
```diff
@@ -848,3 +930,109 @@ describe('ripgrep-search-in-workspace-server', function (): void {
+    });
+});
+
+describe('ripgrep-search-in-workspace-server-filePatternToGlobs', function (): void {
```
```suggestion
describe('#filePatternToGlobs', function (): void {
```

alvsan09 replied:
Done

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +156,94 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+        const args = new Set<string>();
+
+        const appendGlobArgs = (rawPatterns: string[], exclude: boolean) => {
+            rootPaths.forEach(root => {
+                for (const rawPattern of rawPatterns) {
```
nitpick: we might want to use a consistent `for loop` and not mix-match here.

alvsan09 replied:
Good catch!, Done!

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.slow-spec.ts:
```diff
@@ -848,3 +930,109 @@ describe('ripgrep-search-in-workspace-server', function (): void {
+    });
+});
+
+function matchPatternToPathResult(pattern: string, expectedPath: string): void {
```
I'm not sure if the naming can be improved here, the method ultimately checks if the pattern matches the expected path.
It does not `match` the pattern in itself, but rather determines if it matches.

alvsan09 replied:
Renamed to: checkResolvedPathForPattern

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,67 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+        return searchPath;
+    }
+
+    // Attempt to find a path in provided pattern but without a glob suffix
```
nitpick: for consistency

```suggestion
    // Attempt to find a path in provided pattern but without a glob suffix.
```

alvsan09 replied:
Done.

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +156,94 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
     }
 
+    /**
+     * Transforms a file pattern to a glob pattern (unless it's already a glob).
+     * It then creates corresponding 'ripgrep' CLI parameters,
```
It's possible the method does too much, it might be one method to transform the pattern to a glob.
Converting it to CLI arguments seems unrelated.

If that is the intention of the method, it might be better to name it differently.

alvsan09 replied:
Updated the jsdoc and method name to: patternToGlobCLIArguments
this better reflects the purpose of the method.  Thanks!

marechal-p commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -290,6 +400,79 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+    protected resolvePatternToPathMap(patterns: string[], rootPaths: string[], keepSuffix: boolean = false): Map<string, string> {
+        const patternToPathMap = new Map<string, string>();
+
+        patterns.forEach(pattern => {
+            rootPaths.forEach(root => {
+                const foundPath = findPathInSuffixedPattern(root, pattern, keepSuffix);
+
+                if (foundPath) {
+                    patternToPathMap.set(pattern, foundPath);
+                }
+            });
+        });
```
Does that logic mean that when multiple `foundPaths` are found for the same pattern, we'll only keep the last one?

alvsan09 replied:
That's a good catch !
I have updated it so we can now resolve multiple paths for a single pattern, so the resulting map is now <string, string[]>
Done!

alvsan09 commented on packages/search-in-workspace/src/browser/search-in-workspace-result-tree-widget.tsx:
```diff
@@ -517,13 +518,8 @@ export class SearchInWorkspaceResultTreeWidget extends TreeWidget {
 
         // Exclude files already covered by searching open editors.
         this.editorManager.all.forEach(e => {
-            const rootUri = this.workspaceService.getWorkspaceRootUri(e.editor.uri);
-            if (rootUri) {
-                // Exclude pattern beginning with './' works after the fix of #8469.
-                const { name, path } = this.filenameAndPath(e.editor.uri.toString(), rootUri.toString());
-                const excludePath: string = path === '' ? './' + name : path + '/' + name;
-                searchOptions.exclude = (searchOptions.exclude) ? searchOptions.exclude.concat(excludePath) : [excludePath];
-            }
```
Nope, it's best to use the absolute path to the open file in editor, otherwise the relative path could apply to multiple roots and suppress matches inadvertently.

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +41,67 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
 }
 
+/**
+ * Attempts to build an validate an absolute path from a given pattern an a root folder
```
Done

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +159,102 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+
+        const appendGlobArgs = (rawPatterns: string[], exclude: boolean) => {
+            rootPaths.forEach(root => {
+                for (const rawPattern of rawPatterns) {
```
That was done by mistake, thanks for bringing it up !, Done!

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +159,102 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
 
+    /**
+     * Transforms a file pattern to a glob pattern (unless it's already a glob).
+     * It then creates corresponding 'ripgrep' CLI parameters,
```
The jsdoc is now updated!

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +159,102 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+     *
+     * @param rawPattern
+     * @param rootFolder
+     * @param exclude
```
Done,

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -290,6 +411,79 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+     * Transform exclude option patterns from relative paths to absolute paths as long
+     * as the resulting paths get validated.
+     *
+     * @param opts
```
Done

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -290,6 +411,79 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+     * provided as an equivalent search path instead.
+     *
+     * @param rootPaths - Workspace root paths
+     * @param opts
```
Done

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -290,6 +411,79 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+        const patternToPathMap: Map<string, string> = this.resolvePatternToPathMap(opts.exclude, rootPaths, true);
+        const updatedExcludes: string[] = [];
+        opts.exclude.forEach(pattern => {
+            const adjustedPattern = patternToPathMap.get(pattern) ? patternToPathMap.get(pattern) as string : pattern;
```
Done

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -146,7 +256,7 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
         }
 
-        const args = [...rgArgs, what].concat(rootUris.map(root => FileUri.fsPath(root)));
+        const args = [...rgArgs, what].concat(searchPaths);
```
Done

DukeNgn commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +151,98 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+        if (!globArgument.endsWith('*')) {
+            // Add a generic glob CLI argument entry to include files inside a given directory.
+            const suffix = globArgument.endsWith('/') ? '*' : '/*';
+            globArgumentsArray.push(`${globArgument}${suffix}`);
```
Can you explain why it's needed to push `globArgument` to `globArgumentsArray` here?

alvsan09 replied:
A given pattern for include/ exclude files e.g. 'foo' can match a file or directory, the user may be explicit by adding a suffix e.g. '/*' to include files under a corresponding directory, but if the user does not specify the suffix, we add it here so we append two glob arguments e.g. 'foo' and 'foo/*'  this will match a file or files under the given directtory.

DukeNgn commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -290,6 +399,75 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+        const includesAsPaths = this.resolvePatternToPathsMap(opts.include, rootPaths);
+        const patternPaths = Array.from(includesAsPaths.keys());
+
+        // Exclude include file patters that were successfully translated to search paths
```
```suggestion
        // Remove file patterns that were successfully translated to search paths.
```

alvsan09 replied:
Done!

DukeNgn commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -290,6 +399,75 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+        const updatedExcludes: string[] = [];
+        opts.exclude.forEach(pattern => {
+            const maybePath = patternToPathsMap.get(pattern);
+            const adjustedPatterns = maybePath ? maybePath : [pattern];
```
```suggestion
            const adjustedPatterns = maybePath ?? [pattern];
```

alvsan09 replied:
Done!

DukeNgn reviewed (commented):
Just some suggestions 😄
Let me know what you think!

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
 }
 
+/**
+ * Attempts to build an validate an absolute path from a given pattern and a root folder
```
It looks fine, Done !

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+        return searchPath;
+    }
+
+    // Attempt to find a path in provided pattern but without a glob suffix.
```
Done !

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+ */
+function stripPatternGlobSuffix(pattern: string): { patternBase: string, suffix: string | undefined } {
+    let patternBase: string = pattern;
+    const globAllSize: number = 3;
```
Done!

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+
+    // Attempt to find a path in provided pattern but without a glob suffix.
+    let patternBase = undefined;
+    ({ patternBase, suffix } = stripPatternGlobSuffix(pattern));
```
Good to know! Thanks !  Done !

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+    return undefined;
+}
+
+/**
```
It looks fine, Done !

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +151,98 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
-    protected getArgs(options?: SearchInWorkspaceOptions): string[] {
-        const args = ['--hidden', '--json'];
-        args.push(options && options.matchCase ? '--case-sensitive' : '--ignore-case');
+    protected getArgs(rootPaths: string[], options?: SearchInWorkspaceOptions): string[] {
```
Good catch!, it's not longer in used !   I removed it ! Thanks !

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -93,47 +151,98 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+        args.add('--hidden');
+        args.add('--json');
+
+        args.add(options && options.matchCase ? '--case-sensitive' : '--ignore-case');
```
Done!

alvsan09 commented:
Hi @vince-fugnitto , @marechal-p , @DukeNgn, 
I have addressed all comments raised so far, please let me know if there is anything missing.
Regards
/Alvaro

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+ *
+ * Returns the pattern with out the glob stars prefix '/**' or without '\\**'
+ */
+function stripPatternGlobSuffix(pattern: string): { patternBase: string, suffix: string | undefined } {
```
Done

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+        ? pattern.substr(pattern.length - globAllSize, globAllSize) : undefined;
+
+    if (suffix === '/**' || suffix === '\**') {
+        patternBase = pattern.slice(0, -3);
```
Done, although the latest refactoring does not need this variable

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+    const globAllSize: number = 3;
+
+    const suffix: string | undefined = pattern.length > globAllSize
+        ? pattern.substr(pattern.length - globAllSize, globAllSize) : undefined;
```
Done, although the latest commit relies on path.parse instead

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -36,6 +38,62 @@ function bytesOrTextToString(obj: IRgBytesOrText): string {
+    const searchPath = path.join(root, pattern);
+    if (fs.existsSync(searchPath)) {
+        return searchPath;
+    } else if (fs.existsSync(pattern)) {
```
Agree, some left overs after refactoring, Done,

vince-fugnitto reviewed (commented):
@alvsan09 I only have minor comments left regarding the source-code. When addressed can you please squash the commits so that I can verify the final commit message and perform functional tests?

marechal-p commented on packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts:
```diff
@@ -94,45 +95,101 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
+            return `${globCommandArgument}${excludeChar}${updatedPattern}`;
+        };
+
+        const globArgument = toCLIGlobArgument(pattern);
```
Please inline `toCLIGlobArgument` here.

alvsan09 replied:
I can not do it on line 161 above as the resulting globArgument is used in three other places below, 
so I would end up repeating the call... could you please clarify !

marechal-p replied:
```ts
    protected patternToGlobCLIArguments(pattern: string, exclude: boolean): string[] {
        const globCommandArgument = '--glob=';
        const excludeChar = exclude ? '!' : '';
        const subDirGlobPattern = '**/';

        const subDirGlobPrefix = inputPattern.startsWith('/') ? '**' : subDirGlobPattern;
        const updatedPattern = inputPattern.startsWith(subDirGlobPattern) ?
            inputPattern : `${subDirGlobPrefix}${inputPattern}`;

        const globArgument = `${globCommandArgument}${excludeChar}${updatedPattern}`;

        // ...
```

alvsan09 replied:
@marechal-p , the intention of the function is to keep functions smaller to keep the SW readable.
see: https://refactoring.guru/smells/long-method

Having it nested may reduce a bit the sensation of a small function so I have removed it as suggested.
Something to think and discuss in general, so we can apply the best practices on future commits.

marechal-p replied:
To write it down here: Since the `toCLIGlobArgument` was local to the calling function and only ever used one it was useless.

alvsan09 commented on packages/search-in-workspace/src/node/ripgrep-search-paths-resolver.ts:
```diff
@@ -0,0 +1,123 @@
+     *
+     * The second set includes all validated paths derived from joining search paths with patterns.
+     */
+    export function resolvePatternToPaths(patterns: string[], searchPaths: string[]): { convertedPatterns: Set<string>, resolvedPaths: Set<string> } {
```
Done on next commit!

vince-fugnitto commented on packages/search-in-workspace/src/node/ripgrep-search-utils.ts:
```diff
@@ -0,0 +1,123 @@
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import * as fs from '@theia/core/shared/fs-extra';
```
minor: please add a newline.

alvsan09 replied:
Done, Thanks Vince !!

vince-fugnitto reviewed (approved):
I confirmed that the search works for various use-cases :+1: 

- searching like today on master
- prefixing search with `./` for narrower results
- searching outside the workspace (similarly to vscode)
- no visible regressions to search behavior

vince-fugnitto reviewed (approved):
I confirmed that the changes still work well functionally with the latest updates :+1:
